File: python/app/utils/extract_string.py
from app.model.client.ai_model_response import LmStudioContentResponse
import re
import logging

async def extract_thinking_and_context(result: str, prompt_tokens: int = 0, completion_tokens: int = 0, total_tokens: int = 0) -> LmStudioContentResponse:

    logging.info(f"Extracting thinking and context from result: {result}")
    # Extract content between <think> and </think>
    result = result.replace("\n", "")

    return LmStudioContentResponse(
        response=result,
        prompt_tokens=prompt_tokens,
        completion_tokens=completion_tokens,
        total_tokens=total_tokens
    )

def extract_foreign_key_info(input_str: str) -> tuple[list[str], dict[str, str]]:
    """
    Extracts foreign key mapping from the input string.
    Example:
      <linked_field> updated_by, category_id </linked_field>
      <foreign_key_table> users, categories </foreign_key_table>
      <foreign_key_field> id, id </foreign_key_field>
    Returns:
      fields: list of linked fields
      fk_dict: dict mapping each linked field to (foreign_key_table, foreign_key_field)
    """
    import re

    matches = re.findall(r"<([^>]+)>([^<]+)</\1>", input_str)
    if len(matches) != 3:
        raise ValueError("Input must contain exactly three tag pairs.")

    fields = [f.strip() for f in matches[0][1].split(",")]
    tables = [t.strip() for t in matches[1][1].split(",")]
    fk_fields = [fk.strip() for fk in matches[2][1].split(",")]

    if not (len(fields) == len(tables) == len(fk_fields)):
        raise ValueError("Number of fields, tables, and foreign key fields must match.")

    return fields, dict(zip(tables, fk_fields))

def extract_insert_values(insert_sql: str) -> dict:
    """
    Extracts column-value pairs from a single-line SQL INSERT statement.
    Example:
      INSERT INTO tasks (id, title, description, category_id) VALUES ('a5f89c0d-e4b2-46ae-8716-11431ddad3af', 'Design User Interface', 'Create mockups for the new user interface', 'b2e7dcfa-e1ab-460a-8a5a-f9ce555d1234');
      ->
      {
        "id": "a5f89c0d-e4b2-46ae-8716-11431ddad3af",
        "title": "Design User Interface",
        ...
      }
    """
    logging.info(f"Extracting values from SQL: {insert_sql}")

    # Extract columns and values
    match = re.search(
        r"INSERT INTO \w+\s*\(([^)]+)\)\s*VALUES\s*\(([^)]+)\)", insert_sql, re.IGNORECASE
    )
    if not match:
        raise ValueError("Invalid INSERT statement format.")

    columns = [col.strip() for col in match.group(1).split(",")]
    # Split values, handling quoted strings with commas
    values = re.findall(r"'((?:[^']|\\')*)'", match.group(2))

    print(f"Columns: {columns}")
    print(f"Values: {values}")

    if len(columns) != len(values):
        raise ValueError("Number of columns and values do not match.")

    return dict(zip(columns, values))

File: go/utils/extract_string/extract_string.go
package extractstring

import (
	"errors"
	"fmt"
	"log"
	"regexp"
	"strings"
)

type ExtractStringUtils interface{
	ExtractForeignKeyInfo(input string) ([]string, map[string][2]string)
	ExtractInsertValues(sql string) (map[string]string, error)
}

type ExtractStringUtilsImpl struct {
}

func NewExtractStringUtils() ExtractStringUtils {
	return &ExtractStringUtilsImpl{}
}

// ExtractForeignKeyInfo parses XML-style metadata from a string and returns linked fields and a mapping
func (s ExtractStringUtilsImpl) ExtractForeignKeyInfo(input string) ([]string, map[string][2]string) {
	// Helper to extract content inside specific tags
	extractTagContent := func(tag string, input string) []string {
		re := regexp.MustCompile(fmt.Sprintf(`<%s>\s*(.*?)\s*</%s>`, tag, tag))
		matches := re.FindStringSubmatch(input)
		if len(matches) < 2 {
			return []string{}
		}
		parts := strings.Split(matches[1], ",")
		for i := range parts {
			parts[i] = strings.TrimSpace(parts[i])
		}
		return parts
	}

	linkedFields := extractTagContent("linked_field", input)
	foreignTables := extractTagContent("foreign_key_table", input)
	foreignFields := extractTagContent("foreign_key_field", input)

	fkDict := make(map[string][2]string)
	for i := range linkedFields {
		// Only map if all slices are of equal length
		if i < len(foreignTables) && i < len(foreignFields) {
			fkDict[linkedFields[i]] = [2]string{foreignTables[i], foreignFields[i]}
		}
	}

	return linkedFields, fkDict
}

// ExtractInsertValues parses SQL INSERT statement into a map of column-value pairs
func (s ExtractStringUtilsImpl) ExtractInsertValues(sql string) (map[string]string, error) {
	log.Printf("Extracting values from SQL: %s", sql)

	re := regexp.MustCompile(`(?i)INSERT INTO \w+\s*\(([^)]+)\)\s*VALUES\s*\(([^)]+)\)`)
	matches := re.FindStringSubmatch(sql)
	if len(matches) != 3 {
		return nil, errors.New("invalid INSERT statement format")
	}

	columns := strings.Split(matches[1], ",")
	valuesRaw := matches[2]

	// Extract string values including quoted strings with commas
	valueRegex := regexp.MustCompile(`'((?:[^']|\\')*)'`)
	valueMatches := valueRegex.FindAllStringSubmatch(valuesRaw, -1)

	if len(columns) != len(valueMatches) {
		return nil, errors.New("number of columns and values do not match")
	}

	result := make(map[string]string)
	for i, col := range columns {
		result[strings.TrimSpace(col)] = valueMatches[i][1]
	}

	return result, nil
}
